File: dart/glue/lib/src/lib/list/flatten.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Flatten function - flattens nested lists into a single flat list
/// Mirrors Haskell Glue.Lib.List.Flatten.flatten exactly
Ir flatten = IrNativeFunc(flattenImpl);

/// Flatten function implementation
/// Mirrors Haskell Glue.Lib.List.Flatten.flattenImpl exactly
Eval<Ir> flattenImpl(Ir list) {
  return switch (list) {
    IrList(elements: final elements) => flattenList(
      elements.toList(),
    ).map((flattened) => IrList(flattened)),
    _ => throwError(wrongArgumentType(['list'])),
  };
}

/// Helper function to flatten a list recursively
Eval<List<Ir>> flattenList(List<Ir> elements) {
  if (elements.isEmpty) {
    return Eval.pure([]);
  }

  final head = elements[0];
  final tail = elements.sublist(1);

  if (head is IrList) {
    // Head is a list, flatten it and concatenate with flattened tail
    return flattenList(head.elements.toList()).flatMap((flattenedHead) {
      return flattenList(
        tail,
      ).map((flattenedTail) => flattenedHead + flattenedTail);
    });
  } else {
    // Head is not a list, keep it and flatten the tail
    return flattenList(tail).map((flattenedTail) => [head, ...flattenedTail]);
  }
}

File: haskell/glue/src/Glue/Lib/List/Flatten.hs
module Glue.Lib.List.Flatten where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

flatten :: IR Eval
flatten = NativeFunc flattenImpl

flattenImpl :: IR Eval -> Eval (IR Eval)
flattenImpl list = case list of
    List xs -> do
        flattened <- flattenList xs
        pure $ List flattened
    _ -> throwError $ wrongArgumentType ["list"]

-- Helper function to flatten a list recursively
flattenList :: [IR Eval] -> Eval [IR Eval]
flattenList [] = pure []
flattenList (x : xs) = do
    case x of
        List ys -> do
            flattenedHead <- flattenList ys
            flattenedTail <- flattenList xs
            pure $ flattenedHead ++ flattenedTail
        _ -> do
            flattenedTail <- flattenList xs
            pure $ x : flattenedTail
